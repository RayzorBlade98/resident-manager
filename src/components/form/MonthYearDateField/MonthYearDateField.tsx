import { DateField } from '@mui/x-date-pickers';
import dayjs, { Dayjs } from 'dayjs';
import React from 'react';
import '_/extensions/date/date.extension';
import MonthYear from '_/extensions/date/month_year.extension';

interface MonthYearDateFieldProps {
  /**
   * Label of the datepicker field
   */
  label: string;

  /**
   * Current date value of the field
   */
  value: Date | undefined;

  /**
   * Callback when selecting a date
   * @param date selected date
   */
  onChange: (date: MonthYear | undefined) => void;

  /**
   * Error message displayed below the field
   */
  errorMessage?: string;

  /**
   * Id of the date field
   */
  id?: string;

  /**
   * Whether the field is required or not
   */
  required?: boolean;
}

/**
 * Generic date field component
 */
function MonthYearDateField(props: MonthYearDateFieldProps): JSX.Element {
  const onChange = (date: Dayjs | null) => {
    let parsedDate: Date | undefined = date?.toDate();
    if (parsedDate) {
      if (Number.isNaN(parsedDate.valueOf())) {
        parsedDate = undefined;
      } else {
        parsedDate = MonthYear.fromDate(parsedDate.toUTC());
      }
    }
    props.onChange(parsedDate as MonthYear);
  };

  return (
    <DateField
      required={props.required}
      label={props.label}
      format="MM.YYYY"
      defaultValue={props.value ? dayjs(props.value) : undefined}
      onChange={onChange}
      slotProps={{
        textField: {
          helperText: props.errorMessage,
          error: !!props.errorMessage,
          id: props.id,
        },
      }}
    />
  );
}

export default MonthYearDateField;
