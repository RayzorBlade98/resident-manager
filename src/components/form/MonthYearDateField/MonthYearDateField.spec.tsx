import { RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import MonthYearDateField from './MonthYearDateField';
import MonthYear from '_/extensions/date/month_year.extension';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';

describe('MonthYearDateField', () => {
  const onChangeMock = jest.fn();
  const labelText = 'Testlabel';
  let value: MonthYear | undefined;
  let errorMessage: string | undefined;
  let renderResult: RenderResult;

  function renderComponent(): void {
    renderResult = render(
      <RecoilTestWrapper>
        <MonthYearDateField
          id="testField"
          label={labelText}
          value={value}
          onChange={onChangeMock}
          errorMessage={errorMessage}
        />
      </RecoilTestWrapper>,
    );
  }

  beforeEach(() => {
    errorMessage = 'Testerror';
    value = new MonthYear(4, 2023);
    renderComponent();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should display label', () => {
    // Assert
    const label = renderResult.container.querySelector('label');
    expect(label?.textContent).toEqual(labelText);
  });

  test('should display error message', () => {
    // Assert
    const error = renderResult.container.querySelector('p');
    expect(error?.textContent).toEqual(errorMessage);
  });

  test('should display the value', () => {
    // Assert
    const field = renderResult.container.querySelector(
      'input',
    ) as HTMLInputElement;
    expect(field.value).toBe('05.2023');
  });

  test('should call onChange for valid input', async () => {
    // Arrange
    const user = userEvent.setup();
    // Act
    const field = renderResult.getByRole('textbox');
    await user.type(field, '121998');

    // Assert
    expect(onChangeMock).toHaveBeenLastCalledWith(
      new MonthYear(11, 1998).toUTC(),
    );
  });

  test('should call onChange for invalid input', async () => {
    // Arrange
    const user = userEvent.setup();
    // Act
    const field = renderResult.getByRole('textbox');
    await user.click(field);
    await user.keyboard('{Backspace}');

    // Assert
    expect(onChangeMock).toHaveBeenLastCalledWith(undefined);
  });

  test('should match snapshot', () => {
    // Arrange
    errorMessage = undefined;
    renderComponent();

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });

  test('should match snapshot (with error)', () => {
    // Arrange
    value = undefined;
    renderComponent();

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
