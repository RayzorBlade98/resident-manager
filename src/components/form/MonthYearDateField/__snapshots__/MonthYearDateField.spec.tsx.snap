// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthYearDateField should match snapshot (with error) 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="testField"
      id="testField-label"
    >
      Testlabel
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-describedby="testField-helper-text"
        aria-invalid="true"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id="testField"
        inputmode="text"
        placeholder="MM.JJJJ"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Testlabel
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
      id="testField-helper-text"
    >
      Testerror
    </p>
  </div>
</div>
`;

exports[`MonthYearDateField should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="testField"
      id="testField-label"
    >
      Testlabel
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id="testField"
        inputmode="text"
        placeholder="MM.JJJJ"
        type="text"
        value="05.2023"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Testlabel
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
